torch/_dynamo/variables/user_defined.py

```diff
@@ -154,6 +156,55 @@ def call_function(
                     var, "__call_nn_module_init", variables.ConstantVariable(True)
                 )
                 return var
+            elif dataclasses.is_dataclass(self.value):
+                # Manually implement the __init__ function. If not done here,
+                # this goes through a hard to support bytecode for __init__ of
+                # dataclasses.
+                user_cls = self.value
+
+                # Get arg spec
+                keys = [f.name for f in dataclasses.fields(user_cls)]
+                bound = inspect.signature(user_cls).bind(*args, **kwargs)
+                bound.apply_defaults()
+                assert set(bound.arguments.keys()) == set(keys)
+
+                # Get default factory fn
+                init_freevars = user_cls.__init__.__code__.co_freevars
+                init_closure = user_cls.__init__.__closure__ or ()
+                defaults_map = {}
+                for name, cell in zip(init_freevars, init_closure):
+                    defaults_map[name] = cell.cell_contents
+
+                # Build attr, val map
+                items = collections.OrderedDict()
+                for key in keys:
+                    val = bound.arguments[key]
+                    if isinstance(val, VariableTracker):
+                        items[key] = val
+                    elif isinstance(val, dataclasses._HAS_DEFAULT_FACTORY_CLASS):
+                        # This indicates that dataclass has a default factory
+                        # function for a field. We can look at the defaults map
+                        # to find the factory fn.
+                        dataclass_default_prefix = "_dflt_"
+                        default_key = dataclass_default_prefix + key
+                        assert default_key in defaults_map
+                        default_factory_fn = defaults_map[default_key]
+                        if is_builtin_callable(default_factory_fn):
+                            items[key] = variables.BuiltinVariable(
+                                default_factory_fn
+                            ).call_function(tx, (), {})
+                        else:
+                            unimplemented(
+                                "unsupport default factory fn for dataclasses"
+                            )
+                    elif variables.ConstantVariable.is_literal(val):
+                        items[key] = variables.ConstantVariable(val)
+                    else:
+                        unimplemented("Unknown field for dataclass")
+
+                # Force init of the newly created var
+                var.value.__init__(**items)
```

Comment:
This seems wrong.  `var.value` here is just a dummy object whose data should never be accessed.  In the cases above it will just be an empty object with no attributes.  All of the attributes on it should instead be stored as side effects so that if the value is returned we can reconstruct a new instance of the object with all the correct attributes.